common/network-shuffle/src/test/java/org/apache/spark/network/shuffle/RemoteBlockPushResolverSuite.java

```diff
@@ -1022,55 +1019,240 @@ public void testFinalizeShuffleMergeFromPreviousAttemptIsAborted()
+    MergedBlockMeta blockMeta = pushResolver.getMergedBlockMeta(TEST_APP, 0, 2, 0);
+    validateChunks(TEST_APP, 0, 2, 0, blockMeta, new int[]{4}, new int[][]{{0}});
+  }
+
+  @Test
+  public void testFinalizeWithOlderShuffleMergeId() throws IOException {
+    StreamCallbackWithID stream1 =
+      pushResolver.receiveBlockDataAsStream(
+        new PushBlockStream(TEST_APP, NO_ATTEMPT_ID, 0, 1, 0, 0, 0));
+    stream1.onData(stream1.getID(), ByteBuffer.wrap(new byte[2]));
+    StreamCallbackWithID stream2 =
+      pushResolver.receiveBlockDataAsStream(
+        new PushBlockStream(TEST_APP, NO_ATTEMPT_ID, 0, 2, 0, 0, 0));
+    stream2.onData(stream2.getID(), ByteBuffer.wrap(new byte[2]));
+    stream2.onData(stream2.getID(), ByteBuffer.wrap(new byte[2]));
+    stream1.onData(stream1.getID(), ByteBuffer.wrap(new byte[2]));
+    try {
+      // stream 1 push should be rejected as it is from an older shuffleMergeId
+      stream1.onComplete(stream1.getID());
+    } catch(RuntimeException re) {
+      assertEquals("Block shufflePush_0_1_0_0 is received after merged shuffle is finalized or"
+        + " stale block push as shuffle blocks of a higher shuffleMergeId for the shuffle is being"
+          + " pushed", re.getMessage());
+    }
+    // stream 2 now completes
+    stream2.onComplete(stream2.getID());
+    try {
+      pushResolver.finalizeShuffleMerge(new FinalizeShuffleMerge(TEST_APP, NO_ATTEMPT_ID, 0, 1));
+    } catch(RuntimeException re) {
+      assertEquals("Shuffle merge finalize request for shuffle 0 with shuffleMergeId 1 is stale"
+        + " shuffle finalize request as shuffle blocks of a higher shuffleMergeId for the shuffle"
+          + " is already being pushed", re.getMessage());
+    }
+    pushResolver.finalizeShuffleMerge(new FinalizeShuffleMerge(TEST_APP, NO_ATTEMPT_ID, 0, 2));
+
+    MergedBlockMeta blockMeta = pushResolver.getMergedBlockMeta(TEST_APP, 0, 2, 0);
+    validateChunks(TEST_APP, 0, 2, 0, blockMeta, new int[]{4}, new int[][]{{0}});
+  }
+
+  @Test
+  public void testFinalizeOfDeterminateShuffle() throws IOException {
+    PushBlock[] pushBlocks = new PushBlock[] {
+      new PushBlock(0, 0, 0, 0, ByteBuffer.wrap(new byte[4])),
+      new PushBlock(0,  0, 1, 0, ByteBuffer.wrap(new byte[5]))
+    };
+    pushBlockHelper(TEST_APP, NO_ATTEMPT_ID, pushBlocks);
+    MergeStatuses statuses = pushResolver.finalizeShuffleMerge(
+      new FinalizeShuffleMerge(TEST_APP, NO_ATTEMPT_ID, 0, 0));
+
+    RemoteBlockPushResolver.AppShuffleInfo appShuffleInfo =
+      pushResolver.validateAndGetAppShuffleInfo(TEST_APP);
+    assertTrue("Metadata of determinate shuffle should be removed after finalize shuffle"
+      + " merge", appShuffleInfo.getPartitions().get(0) == null);
+    validateMergeStatuses(statuses, new int[] {0}, new long[] {9});
+    MergedBlockMeta blockMeta = pushResolver.getMergedBlockMeta(TEST_APP, 0, 0, 0);
+    validateChunks(TEST_APP, 0, 0, 0, blockMeta, new int[]{4, 5}, new int[][]{{0}, {1}});
```

Comment:
Do we have a test which checks that push to a determinate shuffle after finalize is rejected ?
I think there is one outside of the context of this PR - pls check and add if not.

Reply:
Yes right, I have updated the same test as well `RemoteBlockPushResolverSuite#testBlockReceivedAfterMergeFinalize` checking for the new exception message. Do you think we should add a separate test instead of making changes there?

Reply:
That is sufficient, thanks.